### PR TITLE
Simplify the abortion mechanism with async

### PR DESCRIPTION
**Pull request by pepeiborra**
Proposed fix for #326 
The commit history contains a breakdown of the problem. First I extracted a control flow abstraction, then simplified it using async, and finally inlined it. 

Unfortunately the only thing missing is evidence that this change completely fixes the issue, but hopefully outsourcing the tricky bits to 'async' is a step in the right direction.

### CODE REVIEW

**Comment by ndmitchell** [comment]
Why inline it? Having the abstraction separated seems valuable - and there are a few more things we could hopefully abstract the logic vs control flow.

**Comment by pepeiborra** [comment]
Inlined because the abstraction was essentially Async, and all 3 combinators were async combinators

File: src/Development/IDE/Core/Shake.hs [hunk 1]
```diff
@@ -394,29 +395,28 @@ shakeRun IdeState{shakeExtras=ShakeExtras{..}, ..} acts =
...
-                              in ", profile saved at " <> T.unpack link
-                          _ -> ""
-                  logDebug logger $ T.pack $
-                      "Finishing shakeRun (took " ++ showDuration runTime ++ ", " ++ res' ++ profile ++ ")"
-                  signalBarrier bar (fst <$> res)
-              -- important: we send an async exception to the thread, then wait for it to die, before continuing
-              pure (killThread thread >> void (waitBarrier bar), either throwIO return =<< waitBarrier bar))
+        (\() -> do
+              start <- offsetTime
+              aThread <- asyncWithUnmask $ \restore -> do
```
[the remark this answers is not in the document]
Comment (pepeiborra) [reply src/Development/IDE/Core/Shake.hs]: This is already running in a masked context, isn't it?

[the remark this answers is not in the document]
Comment (cocreature) [reply src/Development/IDE/Core/Shake.hs]: It’s masked but only with `mask` not `uninterruptibleMask`.

[the remark this answers is not in the document]
Comment (pepeiborra) [reply src/Development/IDE/Core/Shake.hs]: I did not know about `uninterruptibleMask`, somehow I had managed to delete that memory from my brain.

[the remark this answers is not in the document]
Comment (pepeiborra) [reply src/Development/IDE/Core/Shake.hs]: The docs at `Control.Exception` say:

> Some operations are interruptible, which means that they can receive asynchronous exceptions even in the scope of a mask

Is `logDebug` interruptible?

[the remark this answers is not in the document]
Comment (pepeiborra) [reply src/Development/IDE/Core/Shake.hs]: I think the sensible thing to do would be to call `logDebug` out of `withMVar`, regardless of whether it is interruptible or not

[the remark this answers is not in the document]
Comment (cocreature) [reply src/Development/IDE/Core/Shake.hs]: Sounds sensible. I think I checked that it’s not interruptible in DAML but the one in `ghcide` looks like it is almost certainly interruptible.

[the remark this answers is not in the document]
Comment (cocreature) [reply src/Development/IDE/Core/Shake.hs]: Looking at the code of `asyncWithUnmask`  it looks like it should be uninterruptible so we should be good now that `logDebug` is moved out.

File: src/Development/IDE/Core/Shake.hs [hunk 2]
```diff
@@ -394,29 +395,31 @@ shakeRun IdeState{shakeExtras=ShakeExtras{..}, ..} acts =
         (\stop -> do
               (stopTime,_) <- duration stop
               logDebug logger $ T.pack $ "Starting shakeRun (aborting the previous one took " ++ showDuration stopTime ++ ")"
-              bar <- newBarrier
-              start <- offsetTime
-              pure (start, bar))
+        )
```
Comment (cocreature) [inline src/Development/IDE/Core/Shake.hs]: I think we should now be able to simplify `withMVar'` since we don’t need to pass anything along.

**Pull request merged**